{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<module 'posixpath' from '/Users/nicolemoiseyev/opt/anaconda3/envs/TCGA/lib/python3.8/posixpath.py'>"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import requests\n",
    "import json\n",
    "import os\n",
    "import sys\n",
    "import re\n",
    "import gzip\n",
    "import shutil\n",
    "import tarfile\n",
    "import pathlib\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from maayanlab_bioinformatics.harmonization import ncbi_genes\n",
    "import math\n",
    "\n",
    "sys.path.append(\"..\") # moved to notebooks directory"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Endpoints\n",
    "base_url = 'https://api.gdc.cancer.gov/'\n",
    "files_endpt = base_url + 'files/'\n",
    "genes_endpt = base_url + 'genes/'\n",
    "cases_endpt = base_url + 'cases/'\n",
    "data_endpt = base_url + \"data/\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def upload_to_aws(local_file, object_name=None):\n",
    "\n",
    "    s3 = boto3.client('s3', aws_access_key_id=config.ACCESS_KEY,\n",
    "                      aws_secret_access_key=config.SECRET_KEY)\n",
    "    \n",
    "    if (object_name == None):\n",
    "        object_name = local_file\n",
    "\n",
    "    s3.upload_file(\n",
    "        local_file, config.BUCKET, object_name\n",
    "    )\n",
    "    \n",
    "# upload_to_aws(\"./data.csv\", \"data.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get info on properties of the returned objects for an endpoint\n",
    "\n",
    "# files:\n",
    "\n",
    "# data = requests.get(files_endpt + \"_mapping\").json()[\"fields\"]\n",
    "# print(json.dumps(data, indent=2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def download_from_aws(object_name, location = None):\n",
    "    \n",
    "    s3 = boto3.client('s3', aws_access_key_id=config.ACCESS_KEY,\n",
    "                      aws_secret_access_key=config.SECRET_KEY)\n",
    "    \n",
    "    if (location == None):\n",
    "        location = \"./\" + object_name\n",
    "    \n",
    "    s3.download_file(\n",
    "        config.BUCKET, object_name, location\n",
    "    )\n",
    "\n",
    "# os.makedirs(\"AWS_test\", exist_ok=True)\n",
    "# download_from_aws(\"data.csv\", \"AWS_test/data.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# RNA-Seq Data\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For now, we'll collect data for one cancer type to save space."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "11280 unique files\n",
      "10393 unique cases\n"
     ]
    }
   ],
   "source": [
    "'''\n",
    "{\n",
    "            \"op\": \"in\",\n",
    "            \"content\":\n",
    "             {\n",
    "                 \"field\": \"cases.diagnoses.primary_diagnosis\", \n",
    "                 \"value\": [],\n",
    "             }\n",
    "         }\n",
    "'''\n",
    "# Retrieve data from the TCGA API\n",
    "\n",
    "# data type of files we want\n",
    "data_type = \"htseq.counts\"\n",
    "\n",
    "# The 'fields' parameter is passed as a comma-separated string of single names.\n",
    "fields = \"file_id,file_name,cases.case_id,cases.diagnoses.primary_diagnosis\"\n",
    "\n",
    "# filter files for only RNA-Seq results\n",
    "filters = {\n",
    "    \"op\": \"and\",\n",
    "     \"content\":[\n",
    "         {\n",
    "            \"op\": \"in\",\n",
    "            \"content\":\n",
    "             {\n",
    "                 \"field\": \"files.experimental_strategy\", \n",
    "                 \"value\": [\"RNA-Seq\"],\n",
    "             }\n",
    "         },\n",
    "         {\n",
    "            \"op\": \"in\",\n",
    "            \"content\":\n",
    "             {\n",
    "                 \"field\": \"access\", \n",
    "                 \"value\": \"open\",\n",
    "                 \n",
    "             }\n",
    "         },\n",
    "         {\n",
    "            \"op\": \"in\",\n",
    "            \"content\":\n",
    "             {\n",
    "                 \"field\": \"files.file_name\", \n",
    "                 \"value\": [\"*htseq.counts.gz\"],\n",
    "             }\n",
    "         },\n",
    "     ],\n",
    "}\n",
    "\n",
    "# build parameters object\n",
    "params = {\n",
    "    \"fields\": fields,\n",
    "    \"filters\": json.dumps(filters),\n",
    "    \"size\": 100000\n",
    "}\n",
    "\n",
    "# get list of all files with RNA-seq results\n",
    "response = requests.get(files_endpt, params = params) # optionally also provide params argument\n",
    "data = json.loads(response.content.decode(\"utf-8\"))\n",
    "\n",
    "# get list of results\n",
    "results = data[\"data\"][\"hits\"]\n",
    "#print(results)\n",
    "#results = list(filter(lambda x: data_type in x[\"file_name\"], results))\n",
    "\n",
    "file_uuid_list = [ entry[\"file_id\"] for entry in results]\n",
    "case_uuid_list = [ entry[\"cases\"][0][\"case_id\"] for entry in results]\n",
    "print(f'{len(file_uuid_list)} unique files')\n",
    "print(f'{len(set(case_uuid_list))} unique cases')\n",
    "\n",
    "\n",
    "df_files_cases=pd.DataFrame({\"case\": case_uuid_list },  index=file_uuid_list)\n",
    "file_to_case = df_files_cases.to_dict()[\"case\"] # dict mapping file id to case id\n",
    "# df_files_cases.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_files_cases.to_csv('files_to_cases.csv', encoding='utf-8')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['Hepatocellular carcinoma, spindle cell variant', 'Pleomorphic carcinoma', 'Adrenal cortical carcinoma', 'Amelanotic melanoma', 'Basaloid squamous cell carcinoma', 'Adenosquamous carcinoma', 'Squamous cell carcinoma, small cell, nonkeratinizing', 'Papillary adenocarcinoma, NOS', 'Superficial spreading melanoma', 'Pheochromocytoma, malignant', 'Mucinous adenocarcinoma, endocervical type', 'Thymoma, type B1, NOS', 'Neuroendocrine carcinoma, NOS', 'Epithelioid mesothelioma, malignant', 'Thymoma, type B2, malignant', 'Medullary carcinoma, NOS', 'Paget disease and infiltrating duct carcinoma of breast', 'Oligodendroglioma, anaplastic', 'Squamous cell carcinoma, keratinizing, NOS', 'Phyllodes tumor, malignant', 'Intraductal micropapillary carcinoma', 'Leiomyosarcoma, NOS', 'Apocrine adenocarcinoma', 'Large cell neuroendocrine carcinoma', 'Papillary serous cystadenocarcinoma', 'Thymoma, type B3, malignant', 'Solid carcinoma, NOS', 'Oligodendroglioma, NOS', 'Mixed epithelioid and spindle cell melanoma', 'Thymoma, type A, NOS', 'Adenocarcinoma with mixed subtypes', 'Renal cell carcinoma, chromophobe type', 'Infiltrating duct mixed with other types of carcinoma', 'Aggressive fibromatosis', 'Papillary carcinoma, NOS', 'Bronchiolo-alveolar carcinoma, non-mucinous', 'Infiltrating duct and lobular carcinoma', 'Pleomorphic liposarcoma', 'Carcinoma, NOS', 'Giant cell sarcoma', 'Cribriform carcinoma, NOS', 'Yolk sac tumor', 'Abdominal fibromatosis', 'Thymoma, type AB, NOS', 'Dedifferentiated liposarcoma', 'Teratoma, malignant, NOS', 'Signet ring cell carcinoma', 'Pheochromocytoma, NOS', 'Synovial sarcoma, biphasic', 'Fibrous mesothelioma, malignant', 'Papillary squamous cell carcinoma', 'Infiltrating duct carcinoma, NOS', 'Myxoid leiomyosarcoma', 'Epithelioid cell melanoma', 'Endometrioid adenocarcinoma, NOS', 'Cholangiocarcinoma', 'Extra-adrenal paraganglioma, malignant', 'Nonencapsulated sclerosing carcinoma', 'Thymoma, type B1, malignant', 'Papillary carcinoma, columnar cell', 'Extra-adrenal paraganglioma, NOS', 'Paraganglioma, NOS', 'Mixed glioma', 'Mesothelioma, malignant', 'Spindle cell melanoma, NOS', 'Malignant melanoma, NOS', 'Seminoma, NOS', 'Liposarcoma, well differentiated', 'Follicular adenocarcinoma, NOS', 'Follicular carcinoma, minimally invasive', 'Squamous cell carcinoma, large cell, nonkeratinizing, NOS', 'Papillary transitional cell carcinoma', 'Spindle cell melanoma, type B', 'Adenocarcinoma, intestinal type', 'Squamous cell carcinoma, NOS', 'Acral lentiginous melanoma, malignant', 'Transitional cell carcinoma', 'Papillary carcinoma, follicular variant', 'Bronchio-alveolar carcinoma, mucinous', 'Nodular melanoma', 'Renal cell carcinoma, NOS', 'Teratocarcinoma', 'Astrocytoma, anaplastic', 'Undifferentiated sarcoma', 'Intraductal papillary adenocarcinoma with invasion', 'Bronchiolo-alveolar adenocarcinoma, NOS', 'Acute myeloid leukemia, NOS', 'Thymoma, type A, malignant', 'Tubular adenocarcinoma', 'Serous cystadenocarcinoma, NOS', 'Desmoplastic melanoma, malignant', 'Carcinosarcoma, NOS', 'Thymoma, type B2, NOS', 'Oxyphilic adenocarcinoma', 'Adenocarcinoma, endocervical type', 'Hepatocellular carcinoma, fibrolamellar', 'Adenocarcinoma with neuroendocrine differentiation', 'Mesodermal mixed tumor', 'Mesothelioma, biphasic, malignant', 'Clear cell adenocarcinoma, NOS', 'Astrocytoma, NOS', 'Glioblastoma', 'Adenoid cystic carcinoma', 'Micropapillary carcinoma, NOS', 'Synovial sarcoma, NOS', 'Adenocarcinoma, NOS', 'Infiltrating lobular mixed with other types of carcinoma', 'Embryonal carcinoma, NOS', 'Carcinoma, diffuse type', 'Hepatocellular carcinoma, clear cell type', 'Squamous cell carcinoma, spindle cell', 'Synovial sarcoma, spindle cell', 'Combined hepatocellular carcinoma and cholangiocarcinoma', 'Endometrioid adenocarcinoma, secretory variant', 'Malignant peripheral nerve sheath tumor', 'Thymoma, type AB, malignant', 'Metaplastic carcinoma, NOS', 'Fibromyxosarcoma', 'Thymic carcinoma, NOS', 'Mullerian mixed tumor', 'Teratoma, benign', 'Carcinoma, undifferentiated, NOS', 'Adenocarcinoma in tubolovillous adenoma', 'Paraganglioma, malignant', 'Acinar cell carcinoma', 'Mucinous adenocarcinoma', 'Secretory carcinoma of breast', 'Lentigo maligna melanoma', 'Malignant lymphoma, large B-cell, diffuse, NOS', 'Serous surface papillary carcinoma', 'Basal cell carcinoma, NOS', 'Malignant fibrous histiocytoma', 'Hepatocellular carcinoma, NOS', 'Lobular carcinoma, NOS', 'Mixed germ cell tumor']\n"
     ]
    }
   ],
   "source": [
    "cancer_types = []\n",
    "\n",
    "for entry in results:\n",
    "    if \"diagnoses\" in entry[\"cases\"][0]:  \n",
    "        cancer_types.append(entry[\"cases\"][0][\"diagnoses\"][0][\"primary_diagnosis\"])\n",
    "    \n",
    "\n",
    "cancer_types = list(set(cancer_types))\n",
    "print(cancer_types)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# generate the manifest file\n",
    "manifest_endpt = \"https://api.gdc.cancer.gov/v0/manifest\"\n",
    "\n",
    "params = {\"ids\": file_uuid_list, \"return_type\": \"manifest\"}\n",
    "\n",
    "\n",
    "response = requests.post(manifest_endpt, data=json.dumps(params), headers={\n",
    "                            \"Content-Type\": \"application/json\"})\n",
    "                        \n",
    "\n",
    "# optionally also provide params argument\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# filename is found in the Content-Disposition header of response\n",
    "response_head_cd = response.headers[\"Content-Disposition\"]\n",
    "file_name = re.findall(\"filename=(.+)\", response_head_cd)[0]\n",
    "\n",
    "downloads_folder = \"manifest_downloads/\"\n",
    "os.makedirs(downloads_folder, exist_ok = True)\n",
    "\n",
    "\n",
    "# Save manifest file to manifest_downloads folder\n",
    "with open(downloads_folder + file_name, \"wb\") as f_out:\n",
    "    f_out.write(response.content)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-9-6bf250f5da04>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;31m# A POST is used, so the filter parameters can be passed directly as a Dict object.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m response = requests.post(data_endpt,\n\u001b[0m\u001b[1;32m      5\u001b[0m                         \u001b[0mdata\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mjson\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdumps\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mparams\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m                         headers={\n",
      "\u001b[0;32m~/opt/anaconda3/envs/TCGA/lib/python3.8/site-packages/requests/api.py\u001b[0m in \u001b[0;36mpost\u001b[0;34m(url, data, json, **kwargs)\u001b[0m\n\u001b[1;32m    117\u001b[0m     \"\"\"\n\u001b[1;32m    118\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 119\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0mrequest\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'post'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0murl\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdata\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mjson\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mjson\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    120\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    121\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/TCGA/lib/python3.8/site-packages/requests/api.py\u001b[0m in \u001b[0;36mrequest\u001b[0;34m(method, url, **kwargs)\u001b[0m\n\u001b[1;32m     59\u001b[0m     \u001b[0;31m# cases, and look like a memory leak in others.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     60\u001b[0m     \u001b[0;32mwith\u001b[0m \u001b[0msessions\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mSession\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0msession\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 61\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0msession\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrequest\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmethod\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mmethod\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0murl\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0murl\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     62\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     63\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/TCGA/lib/python3.8/site-packages/requests/sessions.py\u001b[0m in \u001b[0;36mrequest\u001b[0;34m(self, method, url, params, data, headers, cookies, files, auth, timeout, allow_redirects, proxies, hooks, stream, verify, cert, json)\u001b[0m\n\u001b[1;32m    528\u001b[0m         }\n\u001b[1;32m    529\u001b[0m         \u001b[0msend_kwargs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mupdate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msettings\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 530\u001b[0;31m         \u001b[0mresp\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprep\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0msend_kwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    531\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    532\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mresp\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/TCGA/lib/python3.8/site-packages/requests/sessions.py\u001b[0m in \u001b[0;36msend\u001b[0;34m(self, request, **kwargs)\u001b[0m\n\u001b[1;32m    641\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    642\u001b[0m         \u001b[0;31m# Send the request\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 643\u001b[0;31m         \u001b[0mr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0madapter\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrequest\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    644\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    645\u001b[0m         \u001b[0;31m# Total elapsed time of the request (approximately)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/TCGA/lib/python3.8/site-packages/requests/adapters.py\u001b[0m in \u001b[0;36msend\u001b[0;34m(self, request, stream, timeout, verify, cert, proxies)\u001b[0m\n\u001b[1;32m    437\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    438\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mchunked\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 439\u001b[0;31m                 resp = conn.urlopen(\n\u001b[0m\u001b[1;32m    440\u001b[0m                     \u001b[0mmethod\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mrequest\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmethod\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    441\u001b[0m                     \u001b[0murl\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0murl\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/TCGA/lib/python3.8/site-packages/urllib3/connectionpool.py\u001b[0m in \u001b[0;36murlopen\u001b[0;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)\u001b[0m\n\u001b[1;32m    663\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    664\u001b[0m             \u001b[0;31m# Make the request on the httplib connection object.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 665\u001b[0;31m             httplib_response = self._make_request(\n\u001b[0m\u001b[1;32m    666\u001b[0m                 \u001b[0mconn\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    667\u001b[0m                 \u001b[0mmethod\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/TCGA/lib/python3.8/site-packages/urllib3/connectionpool.py\u001b[0m in \u001b[0;36m_make_request\u001b[0;34m(self, conn, method, url, timeout, chunked, **httplib_request_kw)\u001b[0m\n\u001b[1;32m    419\u001b[0m                     \u001b[0;31m# Python 3 (including for exceptions like SystemExit).\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    420\u001b[0m                     \u001b[0;31m# Otherwise it looks like a bug in the code.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 421\u001b[0;31m                     \u001b[0msix\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mraise_from\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0me\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    422\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mSocketTimeout\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mBaseSSLError\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mSocketError\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    423\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_raise_timeout\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0merr\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0me\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0murl\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0murl\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtimeout_value\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mread_timeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/TCGA/lib/python3.8/site-packages/urllib3/packages/six.py\u001b[0m in \u001b[0;36mraise_from\u001b[0;34m(value, from_value)\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/TCGA/lib/python3.8/site-packages/urllib3/connectionpool.py\u001b[0m in \u001b[0;36m_make_request\u001b[0;34m(self, conn, method, url, timeout, chunked, **httplib_request_kw)\u001b[0m\n\u001b[1;32m    414\u001b[0m                 \u001b[0;31m# Python 3\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    415\u001b[0m                 \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 416\u001b[0;31m                     \u001b[0mhttplib_response\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mconn\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgetresponse\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    417\u001b[0m                 \u001b[0;32mexcept\u001b[0m \u001b[0mBaseException\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    418\u001b[0m                     \u001b[0;31m# Remove the TypeError from the exception chain in\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/TCGA/lib/python3.8/http/client.py\u001b[0m in \u001b[0;36mgetresponse\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   1320\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1321\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1322\u001b[0;31m                 \u001b[0mresponse\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbegin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1323\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mConnectionError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1324\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/TCGA/lib/python3.8/http/client.py\u001b[0m in \u001b[0;36mbegin\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    301\u001b[0m         \u001b[0;31m# read until we get a non-100 response\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    302\u001b[0m         \u001b[0;32mwhile\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 303\u001b[0;31m             \u001b[0mversion\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstatus\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mreason\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_read_status\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    304\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mstatus\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0mCONTINUE\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    305\u001b[0m                 \u001b[0;32mbreak\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/TCGA/lib/python3.8/http/client.py\u001b[0m in \u001b[0;36m_read_status\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    262\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    263\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_read_status\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 264\u001b[0;31m         \u001b[0mline\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreadline\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_MAXLINE\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"iso-8859-1\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    265\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mline\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0m_MAXLINE\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    266\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mLineTooLong\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"status line\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/TCGA/lib/python3.8/socket.py\u001b[0m in \u001b[0;36mreadinto\u001b[0;34m(self, b)\u001b[0m\n\u001b[1;32m    667\u001b[0m         \u001b[0;32mwhile\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    668\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 669\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_sock\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrecv_into\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mb\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    670\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mtimeout\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    671\u001b[0m                 \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_timeout_occurred\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/TCGA/lib/python3.8/site-packages/urllib3/contrib/pyopenssl.py\u001b[0m in \u001b[0;36mrecv_into\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    311\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mrecv_into\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    312\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 313\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconnection\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrecv_into\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    314\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mOpenSSL\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mSSL\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mSysCallError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    315\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msuppress_ragged_eofs\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0margs\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"Unexpected EOF\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/envs/TCGA/lib/python3.8/site-packages/OpenSSL/SSL.py\u001b[0m in \u001b[0;36mrecv_into\u001b[0;34m(self, buffer, nbytes, flags)\u001b[0m\n\u001b[1;32m   1837\u001b[0m             \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_lib\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mSSL_peek\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_ssl\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbuf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnbytes\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1838\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1839\u001b[0;31m             \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_lib\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mSSL_read\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_ssl\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbuf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnbytes\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1840\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_raise_ssl_error\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_ssl\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1841\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "params = {\"ids\": file_uuid_list, \"return_type\": \"manifest\"}\n",
    "\n",
    "# A POST is used, so the filter parameters can be passed directly as a Dict object.\n",
    "response = requests.post(data_endpt,\n",
    "                        data = json.dumps(params),\n",
    "                        headers={\n",
    "                            \"Content-Type\": \"application/json\"})\n",
    "\n",
    "# filename is found in the Content-Disposition header of response\n",
    "response_head_cd = response.headers[\"Content-Disposition\"]\n",
    "file_name = re.findall(\"filename=(.+)\", response_head_cd)[0]\n",
    "\n",
    "downloads_folder = \"TCGA_downloads/\"\n",
    "\n",
    "# Save .tar.gz zipped file to TCGA_downloads folder\n",
    "with open(downloads_folder + file_name, \"wb\") as f_out:\n",
    "    f_out.write(response.content)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "# extract the root tar archive\n",
    "tar = tarfile.open(downloads_folder + file_name, \"r:gz\")\n",
    "tar.extractall(\"./{}\".format(downloads_folder))\n",
    "folder = file_name.split(\".tar.gz\")[0]\n",
    "\n",
    "for tarinfo in tar:\n",
    "    if (tarinfo.name == \"MANIFEST.txt\"): continue\n",
    "    file_id = tarinfo.name.split(\"/\")[0]\n",
    "    \n",
    "    # unzip inner .gz files\n",
    "    with gzip.open(downloads_folder + tarinfo.name, \"rb\") as f_in:\n",
    "        with open(\"data/{}.txt\".format(file_to_case[file_id]), \"wb\") as f_out:\n",
    "            f_out.write(f_in.read())\n",
    "\n",
    "tar.close()\n",
    "\n",
    "'''\n",
    "\n",
    "# initialize empty df\n",
    "df = pd.DataFrame({\"gene\": []})\n",
    "df = df.set_index(\"gene\")\n",
    "\n",
    "size = 0\n",
    "# loop over directories\n",
    "for file_id in os.listdir(\"TCGA_downloads\"):\n",
    "    with gzip.open(f\"TCGA_downloads/{file_id}\", \"rb\") as f_in:\n",
    "        new_df = pd.read_csv(f_in, sep = \"\\t\", header = None)\n",
    "        file_id = re.findall(\"data/(.+).txt\", f_in.name)[0]\n",
    "        new_df.columns = [\"gene\", file_id]\n",
    "        new_df.gene.replace(to_replace = r'\\..*$', value = \"\", regex=True, \n",
    "           inplace=True) # collapse all versions of same gene to one gene\n",
    "        new_df = new_df.set_index(\"gene\")\n",
    "        df = pd.DataFrame.merge(df, new_df, how=\"outer\", left_on = \"gene\", right_on = \"gene\")\n",
    "\n",
    "# drop rows not corresponding to genes        \n",
    "non_genes = list(filter(lambda val: not \"ENSG\" in val,np.array(df.index.values))) \n",
    "df = df.drop(non_genes)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# initialize empty df\n",
    "df = pd.DataFrame({\"gene\": []})\n",
    "df = df.set_index(\"gene\")\n",
    "\n",
    "size = 0\n",
    "# loop over files, merging with pre-existing data\n",
    "for file in pathlib.Path('data').glob('*.txt'):\n",
    "    with open(file, \"rb\") as f_in:\n",
    "        new_df = pd.read_csv(f_in, sep = \"\\t\", header = None)\n",
    "        file_id = re.findall(\"data/(.+).txt\", f_in.name)[0]\n",
    "        new_df.columns = [\"gene\", file_id]\n",
    "        new_df.gene.replace(to_replace = r'\\..*$', value = \"\", regex=True, \n",
    "           inplace=True) # collapse all versions of same gene to one gene\n",
    "        new_df = new_df.set_index(\"gene\")\n",
    "        df = pd.DataFrame.merge(df, new_df, how=\"outer\", left_on = \"gene\", right_on = \"gene\")\n",
    "\n",
    "# drop rows not corresponding to genes        \n",
    "non_genes = list(filter(lambda val: not \"ENSG\" in val,np.array(df.index.values))) \n",
    "df = df.drop(non_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>54ac877a-52ff-450c-9df9-b7cd3fc8e2e2</th>\n",
       "      <th>16e69011-c295-479f-b521-86e66fba498d</th>\n",
       "      <th>b60f22ac-a659-4f33-b01d-820e86a9a5c9</th>\n",
       "      <th>3bce9803-e74c-462e-bd40-8ec337a66f3a</th>\n",
       "      <th>1a094c2b-9fba-4bb1-929d-bb2cdf874bee</th>\n",
       "      <th>2b14123b-8fcd-402c-9399-4e7c47f20252</th>\n",
       "      <th>f58e49fc-9641-4b46-be4c-766445ec70a3</th>\n",
       "      <th>6a1be87b-c4e0-4fd4-b050-50a245b22038</th>\n",
       "      <th>24118b8f-59fe-482a-aed3-7b1bde1571b7</th>\n",
       "      <th>7413c891-f74e-456f-bc70-69c83fb53c70</th>\n",
       "      <th>...</th>\n",
       "      <th>0d23fa5b-95f8-4626-a4d2-c72ad3cc553b</th>\n",
       "      <th>4160e048-f0b0-40f5-805b-e277a5893a3b</th>\n",
       "      <th>0a2d29de-869a-4dc8-ad11-6ee0d0a3a895</th>\n",
       "      <th>b8023162-5e82-40e6-ad8c-8acf81821f01</th>\n",
       "      <th>eec69ea9-ceb8-456a-aff4-41fb8a261e36</th>\n",
       "      <th>5503b3a1-7d03-41b2-9ec8-e478c5414d67</th>\n",
       "      <th>79fd602b-3e8e-4353-aa78-4f5f170b607d</th>\n",
       "      <th>07a859fc-6f78-4905-9035-90e2403dbe8d</th>\n",
       "      <th>f9c835db-2ab6-4bf5-826f-48723493c0ec</th>\n",
       "      <th>a782352e-c4bb-4c3e-ba82-456a47c3689a</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <td>1005</td>\n",
       "      <td>1234</td>\n",
       "      <td>11822</td>\n",
       "      <td>2648</td>\n",
       "      <td>3139</td>\n",
       "      <td>1928</td>\n",
       "      <td>2221</td>\n",
       "      <td>13405</td>\n",
       "      <td>2025</td>\n",
       "      <td>7594</td>\n",
       "      <td>...</td>\n",
       "      <td>3071</td>\n",
       "      <td>4105</td>\n",
       "      <td>3007</td>\n",
       "      <td>6136</td>\n",
       "      <td>8602</td>\n",
       "      <td>4884</td>\n",
       "      <td>2830</td>\n",
       "      <td>5286</td>\n",
       "      <td>3635</td>\n",
       "      <td>4420</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000005</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>7</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <td>232</td>\n",
       "      <td>449</td>\n",
       "      <td>2021</td>\n",
       "      <td>2047</td>\n",
       "      <td>1090</td>\n",
       "      <td>1863</td>\n",
       "      <td>2849</td>\n",
       "      <td>6176</td>\n",
       "      <td>663</td>\n",
       "      <td>3590</td>\n",
       "      <td>...</td>\n",
       "      <td>1399</td>\n",
       "      <td>3706</td>\n",
       "      <td>2297</td>\n",
       "      <td>2954</td>\n",
       "      <td>3094</td>\n",
       "      <td>5054</td>\n",
       "      <td>1045</td>\n",
       "      <td>3846</td>\n",
       "      <td>5822</td>\n",
       "      <td>4038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457</th>\n",
       "      <td>115</td>\n",
       "      <td>356</td>\n",
       "      <td>1157</td>\n",
       "      <td>453</td>\n",
       "      <td>404</td>\n",
       "      <td>651</td>\n",
       "      <td>695</td>\n",
       "      <td>1060</td>\n",
       "      <td>315</td>\n",
       "      <td>1410</td>\n",
       "      <td>...</td>\n",
       "      <td>759</td>\n",
       "      <td>653</td>\n",
       "      <td>975</td>\n",
       "      <td>1194</td>\n",
       "      <td>1407</td>\n",
       "      <td>890</td>\n",
       "      <td>718</td>\n",
       "      <td>981</td>\n",
       "      <td>1033</td>\n",
       "      <td>978</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <td>10</td>\n",
       "      <td>209</td>\n",
       "      <td>1421</td>\n",
       "      <td>412</td>\n",
       "      <td>209</td>\n",
       "      <td>537</td>\n",
       "      <td>448</td>\n",
       "      <td>693</td>\n",
       "      <td>226</td>\n",
       "      <td>634</td>\n",
       "      <td>...</td>\n",
       "      <td>201</td>\n",
       "      <td>419</td>\n",
       "      <td>777</td>\n",
       "      <td>740</td>\n",
       "      <td>1525</td>\n",
       "      <td>626</td>\n",
       "      <td>198</td>\n",
       "      <td>753</td>\n",
       "      <td>915</td>\n",
       "      <td>724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000938</th>\n",
       "      <td>17</td>\n",
       "      <td>994</td>\n",
       "      <td>93</td>\n",
       "      <td>1030</td>\n",
       "      <td>32</td>\n",
       "      <td>489</td>\n",
       "      <td>386</td>\n",
       "      <td>298</td>\n",
       "      <td>149</td>\n",
       "      <td>1055</td>\n",
       "      <td>...</td>\n",
       "      <td>94</td>\n",
       "      <td>224</td>\n",
       "      <td>684</td>\n",
       "      <td>127</td>\n",
       "      <td>1226</td>\n",
       "      <td>1093</td>\n",
       "      <td>332</td>\n",
       "      <td>527</td>\n",
       "      <td>287</td>\n",
       "      <td>477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000971</th>\n",
       "      <td>491</td>\n",
       "      <td>5369</td>\n",
       "      <td>788</td>\n",
       "      <td>1797</td>\n",
       "      <td>166</td>\n",
       "      <td>2699</td>\n",
       "      <td>1604</td>\n",
       "      <td>8482</td>\n",
       "      <td>393</td>\n",
       "      <td>727</td>\n",
       "      <td>...</td>\n",
       "      <td>7588</td>\n",
       "      <td>878</td>\n",
       "      <td>2083</td>\n",
       "      <td>405</td>\n",
       "      <td>3430</td>\n",
       "      <td>9918</td>\n",
       "      <td>5980</td>\n",
       "      <td>4381</td>\n",
       "      <td>2322</td>\n",
       "      <td>2536</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000001036</th>\n",
       "      <td>1105</td>\n",
       "      <td>2574</td>\n",
       "      <td>3808</td>\n",
       "      <td>4219</td>\n",
       "      <td>804</td>\n",
       "      <td>3926</td>\n",
       "      <td>2715</td>\n",
       "      <td>4154</td>\n",
       "      <td>1374</td>\n",
       "      <td>3310</td>\n",
       "      <td>...</td>\n",
       "      <td>1880</td>\n",
       "      <td>3021</td>\n",
       "      <td>2682</td>\n",
       "      <td>5720</td>\n",
       "      <td>6115</td>\n",
       "      <td>4445</td>\n",
       "      <td>2319</td>\n",
       "      <td>4660</td>\n",
       "      <td>2768</td>\n",
       "      <td>3301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000001084</th>\n",
       "      <td>229</td>\n",
       "      <td>893</td>\n",
       "      <td>1914</td>\n",
       "      <td>1436</td>\n",
       "      <td>384</td>\n",
       "      <td>1665</td>\n",
       "      <td>1782</td>\n",
       "      <td>2435</td>\n",
       "      <td>805</td>\n",
       "      <td>2761</td>\n",
       "      <td>...</td>\n",
       "      <td>1649</td>\n",
       "      <td>1909</td>\n",
       "      <td>1200</td>\n",
       "      <td>4457</td>\n",
       "      <td>3229</td>\n",
       "      <td>2070</td>\n",
       "      <td>1446</td>\n",
       "      <td>2462</td>\n",
       "      <td>2159</td>\n",
       "      <td>2292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000001167</th>\n",
       "      <td>433</td>\n",
       "      <td>622</td>\n",
       "      <td>4450</td>\n",
       "      <td>1963</td>\n",
       "      <td>1165</td>\n",
       "      <td>1882</td>\n",
       "      <td>2162</td>\n",
       "      <td>1938</td>\n",
       "      <td>868</td>\n",
       "      <td>2954</td>\n",
       "      <td>...</td>\n",
       "      <td>570</td>\n",
       "      <td>1573</td>\n",
       "      <td>1391</td>\n",
       "      <td>3964</td>\n",
       "      <td>4094</td>\n",
       "      <td>2381</td>\n",
       "      <td>1023</td>\n",
       "      <td>2272</td>\n",
       "      <td>1498</td>\n",
       "      <td>2319</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10 rows Ã— 506 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 54ac877a-52ff-450c-9df9-b7cd3fc8e2e2  \\\n",
       "gene                                                    \n",
       "ENSG00000000003                                  1005   \n",
       "ENSG00000000005                                     2   \n",
       "ENSG00000000419                                   232   \n",
       "ENSG00000000457                                   115   \n",
       "ENSG00000000460                                    10   \n",
       "ENSG00000000938                                    17   \n",
       "ENSG00000000971                                   491   \n",
       "ENSG00000001036                                  1105   \n",
       "ENSG00000001084                                   229   \n",
       "ENSG00000001167                                   433   \n",
       "\n",
       "                 16e69011-c295-479f-b521-86e66fba498d  \\\n",
       "gene                                                    \n",
       "ENSG00000000003                                  1234   \n",
       "ENSG00000000005                                     0   \n",
       "ENSG00000000419                                   449   \n",
       "ENSG00000000457                                   356   \n",
       "ENSG00000000460                                   209   \n",
       "ENSG00000000938                                   994   \n",
       "ENSG00000000971                                  5369   \n",
       "ENSG00000001036                                  2574   \n",
       "ENSG00000001084                                   893   \n",
       "ENSG00000001167                                   622   \n",
       "\n",
       "                 b60f22ac-a659-4f33-b01d-820e86a9a5c9  \\\n",
       "gene                                                    \n",
       "ENSG00000000003                                 11822   \n",
       "ENSG00000000005                                    10   \n",
       "ENSG00000000419                                  2021   \n",
       "ENSG00000000457                                  1157   \n",
       "ENSG00000000460                                  1421   \n",
       "ENSG00000000938                                    93   \n",
       "ENSG00000000971                                   788   \n",
       "ENSG00000001036                                  3808   \n",
       "ENSG00000001084                                  1914   \n",
       "ENSG00000001167                                  4450   \n",
       "\n",
       "                 3bce9803-e74c-462e-bd40-8ec337a66f3a  \\\n",
       "gene                                                    \n",
       "ENSG00000000003                                  2648   \n",
       "ENSG00000000005                                     7   \n",
       "ENSG00000000419                                  2047   \n",
       "ENSG00000000457                                   453   \n",
       "ENSG00000000460                                   412   \n",
       "ENSG00000000938                                  1030   \n",
       "ENSG00000000971                                  1797   \n",
       "ENSG00000001036                                  4219   \n",
       "ENSG00000001084                                  1436   \n",
       "ENSG00000001167                                  1963   \n",
       "\n",
       "                 1a094c2b-9fba-4bb1-929d-bb2cdf874bee  \\\n",
       "gene                                                    \n",
       "ENSG00000000003                                  3139   \n",
       "ENSG00000000005                                     4   \n",
       "ENSG00000000419                                  1090   \n",
       "ENSG00000000457                                   404   \n",
       "ENSG00000000460                                   209   \n",
       "ENSG00000000938                                    32   \n",
       "ENSG00000000971                                   166   \n",
       "ENSG00000001036                                   804   \n",
       "ENSG00000001084                                   384   \n",
       "ENSG00000001167                                  1165   \n",
       "\n",
       "                 2b14123b-8fcd-402c-9399-4e7c47f20252  \\\n",
       "gene                                                    \n",
       "ENSG00000000003                                  1928   \n",
       "ENSG00000000005                                     3   \n",
       "ENSG00000000419                                  1863   \n",
       "ENSG00000000457                                   651   \n",
       "ENSG00000000460                                   537   \n",
       "ENSG00000000938                                   489   \n",
       "ENSG00000000971                                  2699   \n",
       "ENSG00000001036                                  3926   \n",
       "ENSG00000001084                                  1665   \n",
       "ENSG00000001167                                  1882   \n",
       "\n",
       "                 f58e49fc-9641-4b46-be4c-766445ec70a3  \\\n",
       "gene                                                    \n",
       "ENSG00000000003                                  2221   \n",
       "ENSG00000000005                                     1   \n",
       "ENSG00000000419                                  2849   \n",
       "ENSG00000000457                                   695   \n",
       "ENSG00000000460                                   448   \n",
       "ENSG00000000938                                   386   \n",
       "ENSG00000000971                                  1604   \n",
       "ENSG00000001036                                  2715   \n",
       "ENSG00000001084                                  1782   \n",
       "ENSG00000001167                                  2162   \n",
       "\n",
       "                 6a1be87b-c4e0-4fd4-b050-50a245b22038  \\\n",
       "gene                                                    \n",
       "ENSG00000000003                                 13405   \n",
       "ENSG00000000005                                     2   \n",
       "ENSG00000000419                                  6176   \n",
       "ENSG00000000457                                  1060   \n",
       "ENSG00000000460                                   693   \n",
       "ENSG00000000938                                   298   \n",
       "ENSG00000000971                                  8482   \n",
       "ENSG00000001036                                  4154   \n",
       "ENSG00000001084                                  2435   \n",
       "ENSG00000001167                                  1938   \n",
       "\n",
       "                 24118b8f-59fe-482a-aed3-7b1bde1571b7  \\\n",
       "gene                                                    \n",
       "ENSG00000000003                                  2025   \n",
       "ENSG00000000005                                     2   \n",
       "ENSG00000000419                                   663   \n",
       "ENSG00000000457                                   315   \n",
       "ENSG00000000460                                   226   \n",
       "ENSG00000000938                                   149   \n",
       "ENSG00000000971                                   393   \n",
       "ENSG00000001036                                  1374   \n",
       "ENSG00000001084                                   805   \n",
       "ENSG00000001167                                   868   \n",
       "\n",
       "                 7413c891-f74e-456f-bc70-69c83fb53c70  ...  \\\n",
       "gene                                                   ...   \n",
       "ENSG00000000003                                  7594  ...   \n",
       "ENSG00000000005                                     6  ...   \n",
       "ENSG00000000419                                  3590  ...   \n",
       "ENSG00000000457                                  1410  ...   \n",
       "ENSG00000000460                                   634  ...   \n",
       "ENSG00000000938                                  1055  ...   \n",
       "ENSG00000000971                                   727  ...   \n",
       "ENSG00000001036                                  3310  ...   \n",
       "ENSG00000001084                                  2761  ...   \n",
       "ENSG00000001167                                  2954  ...   \n",
       "\n",
       "                 0d23fa5b-95f8-4626-a4d2-c72ad3cc553b  \\\n",
       "gene                                                    \n",
       "ENSG00000000003                                  3071   \n",
       "ENSG00000000005                                     2   \n",
       "ENSG00000000419                                  1399   \n",
       "ENSG00000000457                                   759   \n",
       "ENSG00000000460                                   201   \n",
       "ENSG00000000938                                    94   \n",
       "ENSG00000000971                                  7588   \n",
       "ENSG00000001036                                  1880   \n",
       "ENSG00000001084                                  1649   \n",
       "ENSG00000001167                                   570   \n",
       "\n",
       "                 4160e048-f0b0-40f5-805b-e277a5893a3b  \\\n",
       "gene                                                    \n",
       "ENSG00000000003                                  4105   \n",
       "ENSG00000000005                                     3   \n",
       "ENSG00000000419                                  3706   \n",
       "ENSG00000000457                                   653   \n",
       "ENSG00000000460                                   419   \n",
       "ENSG00000000938                                   224   \n",
       "ENSG00000000971                                   878   \n",
       "ENSG00000001036                                  3021   \n",
       "ENSG00000001084                                  1909   \n",
       "ENSG00000001167                                  1573   \n",
       "\n",
       "                 0a2d29de-869a-4dc8-ad11-6ee0d0a3a895  \\\n",
       "gene                                                    \n",
       "ENSG00000000003                                  3007   \n",
       "ENSG00000000005                                     0   \n",
       "ENSG00000000419                                  2297   \n",
       "ENSG00000000457                                   975   \n",
       "ENSG00000000460                                   777   \n",
       "ENSG00000000938                                   684   \n",
       "ENSG00000000971                                  2083   \n",
       "ENSG00000001036                                  2682   \n",
       "ENSG00000001084                                  1200   \n",
       "ENSG00000001167                                  1391   \n",
       "\n",
       "                 b8023162-5e82-40e6-ad8c-8acf81821f01  \\\n",
       "gene                                                    \n",
       "ENSG00000000003                                  6136   \n",
       "ENSG00000000005                                     8   \n",
       "ENSG00000000419                                  2954   \n",
       "ENSG00000000457                                  1194   \n",
       "ENSG00000000460                                   740   \n",
       "ENSG00000000938                                   127   \n",
       "ENSG00000000971                                   405   \n",
       "ENSG00000001036                                  5720   \n",
       "ENSG00000001084                                  4457   \n",
       "ENSG00000001167                                  3964   \n",
       "\n",
       "                 eec69ea9-ceb8-456a-aff4-41fb8a261e36  \\\n",
       "gene                                                    \n",
       "ENSG00000000003                                  8602   \n",
       "ENSG00000000005                                    32   \n",
       "ENSG00000000419                                  3094   \n",
       "ENSG00000000457                                  1407   \n",
       "ENSG00000000460                                  1525   \n",
       "ENSG00000000938                                  1226   \n",
       "ENSG00000000971                                  3430   \n",
       "ENSG00000001036                                  6115   \n",
       "ENSG00000001084                                  3229   \n",
       "ENSG00000001167                                  4094   \n",
       "\n",
       "                 5503b3a1-7d03-41b2-9ec8-e478c5414d67  \\\n",
       "gene                                                    \n",
       "ENSG00000000003                                  4884   \n",
       "ENSG00000000005                                     2   \n",
       "ENSG00000000419                                  5054   \n",
       "ENSG00000000457                                   890   \n",
       "ENSG00000000460                                   626   \n",
       "ENSG00000000938                                  1093   \n",
       "ENSG00000000971                                  9918   \n",
       "ENSG00000001036                                  4445   \n",
       "ENSG00000001084                                  2070   \n",
       "ENSG00000001167                                  2381   \n",
       "\n",
       "                 79fd602b-3e8e-4353-aa78-4f5f170b607d  \\\n",
       "gene                                                    \n",
       "ENSG00000000003                                  2830   \n",
       "ENSG00000000005                                     3   \n",
       "ENSG00000000419                                  1045   \n",
       "ENSG00000000457                                   718   \n",
       "ENSG00000000460                                   198   \n",
       "ENSG00000000938                                   332   \n",
       "ENSG00000000971                                  5980   \n",
       "ENSG00000001036                                  2319   \n",
       "ENSG00000001084                                  1446   \n",
       "ENSG00000001167                                  1023   \n",
       "\n",
       "                 07a859fc-6f78-4905-9035-90e2403dbe8d  \\\n",
       "gene                                                    \n",
       "ENSG00000000003                                  5286   \n",
       "ENSG00000000005                                     4   \n",
       "ENSG00000000419                                  3846   \n",
       "ENSG00000000457                                   981   \n",
       "ENSG00000000460                                   753   \n",
       "ENSG00000000938                                   527   \n",
       "ENSG00000000971                                  4381   \n",
       "ENSG00000001036                                  4660   \n",
       "ENSG00000001084                                  2462   \n",
       "ENSG00000001167                                  2272   \n",
       "\n",
       "                 f9c835db-2ab6-4bf5-826f-48723493c0ec  \\\n",
       "gene                                                    \n",
       "ENSG00000000003                                  3635   \n",
       "ENSG00000000005                                     3   \n",
       "ENSG00000000419                                  5822   \n",
       "ENSG00000000457                                  1033   \n",
       "ENSG00000000460                                   915   \n",
       "ENSG00000000938                                   287   \n",
       "ENSG00000000971                                  2322   \n",
       "ENSG00000001036                                  2768   \n",
       "ENSG00000001084                                  2159   \n",
       "ENSG00000001167                                  1498   \n",
       "\n",
       "                 a782352e-c4bb-4c3e-ba82-456a47c3689a  \n",
       "gene                                                   \n",
       "ENSG00000000003                                  4420  \n",
       "ENSG00000000005                                     0  \n",
       "ENSG00000000419                                  4038  \n",
       "ENSG00000000457                                   978  \n",
       "ENSG00000000460                                   724  \n",
       "ENSG00000000938                                   477  \n",
       "ENSG00000000971                                  2536  \n",
       "ENSG00000001036                                  3301  \n",
       "ENSG00000001084                                  2292  \n",
       "ENSG00000001167                                  2319  \n",
       "\n",
       "[10 rows x 506 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# API only accepts 1000 ids per call\n",
    "\n",
    "# Convert Ensembl gene IDs to Entrez gene symbols\n",
    "# mygene_endpt = \"http://mygene.info/v3/gene/\"\n",
    "# headers = {'content-type': 'application/x-www-form-urlencoded'}\n",
    "\n",
    "# ensembl_ids = df.index.to_list()\n",
    "# build parameters object\n",
    "\n",
    "# ids = \",\".join(ensembl_ids)\n",
    "# species = \"human\"\n",
    "# fields = \",\".join([\"symbol\",\"name\",\"taxid\",\"entrezgene\"])\n",
    "\n",
    "# params = \"ids={}&species={}&fields={}\".format(ids,species,fields)\n",
    "# response = requests.post(mygene_endpt, data=params, headers=headers)\n",
    "# data = response.json()\n",
    "\n",
    "# create map of ensembl ID -> entrez symbol\n",
    "# ensembl_to_symbol = {gene[\"query\"]:gene[\"symbol\"] for gene in data}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Lots of missing values\n",
    "# id_chart = pd.read_csv(\"gene_id_conversion.txt\", sep=\"\\t\")\n",
    "# id_chart = id_chart[[\"Ensembl gene ID\", \"Approved symbol\"]]\n",
    "# id_chart = id_chart.set_index(\"Ensembl gene ID\")\n",
    "# ensembl_to_symbol = id_chart.to_dict()[\"Approved symbol\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/nicolemoiseyev/opt/anaconda3/envs/TCGA/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3331: DtypeWarning: Columns (14) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  exec(code_obj, self.user_global_ns, self.user_ns)\n"
     ]
    }
   ],
   "source": [
    "ncbi = pd.DataFrame(ncbi_genes.ncbi_genes_fetch())\n",
    "\n",
    "def get_ensembl_id(ids):\n",
    "    ids = \"\".join(ids)\n",
    "    ensembl = re.findall(\"Ensembl:(.*)\", ids)\n",
    "    if (len(ensembl) == 1):\n",
    "        return ensembl[0]\n",
    "    else:\n",
    "        return None\n",
    "\n",
    "    \n",
    "all_ids = ncbi.dbXrefs.values\n",
    "ensembl_ids = [ get_ensembl_id(ids) for ids in all_ids] \n",
    "\n",
    "ncbi = ncbi[[\"dbXrefs\", \"Symbol\", \"type_of_gene\"]]\n",
    "ncbi[\"ensembl\"] = ensembl_ids\n",
    "ncbi = ncbi.drop(columns=[\"dbXrefs\"])\n",
    "ncbi = ncbi.set_index(\"ensembl\")\n",
    "\n",
    "ensembl_to_symbol = ncbi.to_dict()[\"Symbol\"]\n",
    "ensembl_to_gene_type = ncbi.to_dict()[\"type_of_gene\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_ensembl_ids = df.index.to_list()\n",
    "\n",
    "# if the key is present, return it; otherwise, set the index for the corresponding row as its ensembl id\n",
    "def id_to_symbol(key):\n",
    "    if (key in ensembl_to_symbol): \n",
    "        return ensembl_to_symbol[key]\n",
    "    else:\n",
    "        return key # can return key here to maintain some gene identity\n",
    "\n",
    "def id_to_type(key):\n",
    "    if (key in ensembl_to_gene_type): \n",
    "        return ensembl_to_gene_type[key]\n",
    "    else:\n",
    "        return None\n",
    "    \n",
    "data_symbols = [ id_to_symbol(key) for key in data_ensembl_ids ]\n",
    "data_types = [ id_to_type(key) for key in data_ensembl_ids ]\n",
    "\n",
    "df[\"symbol\"] = data_symbols\n",
    "df[\"type_of_gene\"] = data_types\n",
    "df_symbol = df.set_index(\"symbol\")\n",
    "\n",
    "# drop non protein-coding genes\n",
    "df_symbol = df_symbol[df_symbol[\"type_of_gene\"] == \"protein-coding\"]\n",
    "df_symbol = df_symbol.drop(columns=[\"type_of_gene\"])\n",
    "df_symbol.head()\n",
    "\n",
    "# write the final csv, ready for normalization and further preprocessing\n",
    "df_symbol.to_csv('data.csv', encoding='utf-8')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>symbol</th>\n",
       "      <th>54ac877a-52ff-450c-9df9-b7cd3fc8e2e2</th>\n",
       "      <th>16e69011-c295-479f-b521-86e66fba498d</th>\n",
       "      <th>b60f22ac-a659-4f33-b01d-820e86a9a5c9</th>\n",
       "      <th>3bce9803-e74c-462e-bd40-8ec337a66f3a</th>\n",
       "      <th>1a094c2b-9fba-4bb1-929d-bb2cdf874bee</th>\n",
       "      <th>2b14123b-8fcd-402c-9399-4e7c47f20252</th>\n",
       "      <th>f58e49fc-9641-4b46-be4c-766445ec70a3</th>\n",
       "      <th>6a1be87b-c4e0-4fd4-b050-50a245b22038</th>\n",
       "      <th>24118b8f-59fe-482a-aed3-7b1bde1571b7</th>\n",
       "      <th>...</th>\n",
       "      <th>0d23fa5b-95f8-4626-a4d2-c72ad3cc553b</th>\n",
       "      <th>4160e048-f0b0-40f5-805b-e277a5893a3b</th>\n",
       "      <th>0a2d29de-869a-4dc8-ad11-6ee0d0a3a895</th>\n",
       "      <th>b8023162-5e82-40e6-ad8c-8acf81821f01</th>\n",
       "      <th>eec69ea9-ceb8-456a-aff4-41fb8a261e36</th>\n",
       "      <th>5503b3a1-7d03-41b2-9ec8-e478c5414d67</th>\n",
       "      <th>79fd602b-3e8e-4353-aa78-4f5f170b607d</th>\n",
       "      <th>07a859fc-6f78-4905-9035-90e2403dbe8d</th>\n",
       "      <th>f9c835db-2ab6-4bf5-826f-48723493c0ec</th>\n",
       "      <th>a782352e-c4bb-4c3e-ba82-456a47c3689a</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>1005</td>\n",
       "      <td>1234</td>\n",
       "      <td>11822</td>\n",
       "      <td>2648</td>\n",
       "      <td>3139</td>\n",
       "      <td>1928</td>\n",
       "      <td>2221</td>\n",
       "      <td>13405</td>\n",
       "      <td>2025</td>\n",
       "      <td>...</td>\n",
       "      <td>3071</td>\n",
       "      <td>4105</td>\n",
       "      <td>3007</td>\n",
       "      <td>6136</td>\n",
       "      <td>8602</td>\n",
       "      <td>4884</td>\n",
       "      <td>2830</td>\n",
       "      <td>5286</td>\n",
       "      <td>3635</td>\n",
       "      <td>4420</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TNMD</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>7</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>32</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DPM1</td>\n",
       "      <td>232</td>\n",
       "      <td>449</td>\n",
       "      <td>2021</td>\n",
       "      <td>2047</td>\n",
       "      <td>1090</td>\n",
       "      <td>1863</td>\n",
       "      <td>2849</td>\n",
       "      <td>6176</td>\n",
       "      <td>663</td>\n",
       "      <td>...</td>\n",
       "      <td>1399</td>\n",
       "      <td>3706</td>\n",
       "      <td>2297</td>\n",
       "      <td>2954</td>\n",
       "      <td>3094</td>\n",
       "      <td>5054</td>\n",
       "      <td>1045</td>\n",
       "      <td>3846</td>\n",
       "      <td>5822</td>\n",
       "      <td>4038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SCYL3</td>\n",
       "      <td>115</td>\n",
       "      <td>356</td>\n",
       "      <td>1157</td>\n",
       "      <td>453</td>\n",
       "      <td>404</td>\n",
       "      <td>651</td>\n",
       "      <td>695</td>\n",
       "      <td>1060</td>\n",
       "      <td>315</td>\n",
       "      <td>...</td>\n",
       "      <td>759</td>\n",
       "      <td>653</td>\n",
       "      <td>975</td>\n",
       "      <td>1194</td>\n",
       "      <td>1407</td>\n",
       "      <td>890</td>\n",
       "      <td>718</td>\n",
       "      <td>981</td>\n",
       "      <td>1033</td>\n",
       "      <td>978</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>C1orf112</td>\n",
       "      <td>10</td>\n",
       "      <td>209</td>\n",
       "      <td>1421</td>\n",
       "      <td>412</td>\n",
       "      <td>209</td>\n",
       "      <td>537</td>\n",
       "      <td>448</td>\n",
       "      <td>693</td>\n",
       "      <td>226</td>\n",
       "      <td>...</td>\n",
       "      <td>201</td>\n",
       "      <td>419</td>\n",
       "      <td>777</td>\n",
       "      <td>740</td>\n",
       "      <td>1525</td>\n",
       "      <td>626</td>\n",
       "      <td>198</td>\n",
       "      <td>753</td>\n",
       "      <td>915</td>\n",
       "      <td>724</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 507 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     symbol  54ac877a-52ff-450c-9df9-b7cd3fc8e2e2  \\\n",
       "0    TSPAN6                                  1005   \n",
       "1      TNMD                                     2   \n",
       "2      DPM1                                   232   \n",
       "3     SCYL3                                   115   \n",
       "4  C1orf112                                    10   \n",
       "\n",
       "   16e69011-c295-479f-b521-86e66fba498d  b60f22ac-a659-4f33-b01d-820e86a9a5c9  \\\n",
       "0                                  1234                                 11822   \n",
       "1                                     0                                    10   \n",
       "2                                   449                                  2021   \n",
       "3                                   356                                  1157   \n",
       "4                                   209                                  1421   \n",
       "\n",
       "   3bce9803-e74c-462e-bd40-8ec337a66f3a  1a094c2b-9fba-4bb1-929d-bb2cdf874bee  \\\n",
       "0                                  2648                                  3139   \n",
       "1                                     7                                     4   \n",
       "2                                  2047                                  1090   \n",
       "3                                   453                                   404   \n",
       "4                                   412                                   209   \n",
       "\n",
       "   2b14123b-8fcd-402c-9399-4e7c47f20252  f58e49fc-9641-4b46-be4c-766445ec70a3  \\\n",
       "0                                  1928                                  2221   \n",
       "1                                     3                                     1   \n",
       "2                                  1863                                  2849   \n",
       "3                                   651                                   695   \n",
       "4                                   537                                   448   \n",
       "\n",
       "   6a1be87b-c4e0-4fd4-b050-50a245b22038  24118b8f-59fe-482a-aed3-7b1bde1571b7  \\\n",
       "0                                 13405                                  2025   \n",
       "1                                     2                                     2   \n",
       "2                                  6176                                   663   \n",
       "3                                  1060                                   315   \n",
       "4                                   693                                   226   \n",
       "\n",
       "   ...  0d23fa5b-95f8-4626-a4d2-c72ad3cc553b  \\\n",
       "0  ...                                  3071   \n",
       "1  ...                                     2   \n",
       "2  ...                                  1399   \n",
       "3  ...                                   759   \n",
       "4  ...                                   201   \n",
       "\n",
       "   4160e048-f0b0-40f5-805b-e277a5893a3b  0a2d29de-869a-4dc8-ad11-6ee0d0a3a895  \\\n",
       "0                                  4105                                  3007   \n",
       "1                                     3                                     0   \n",
       "2                                  3706                                  2297   \n",
       "3                                   653                                   975   \n",
       "4                                   419                                   777   \n",
       "\n",
       "   b8023162-5e82-40e6-ad8c-8acf81821f01  eec69ea9-ceb8-456a-aff4-41fb8a261e36  \\\n",
       "0                                  6136                                  8602   \n",
       "1                                     8                                    32   \n",
       "2                                  2954                                  3094   \n",
       "3                                  1194                                  1407   \n",
       "4                                   740                                  1525   \n",
       "\n",
       "   5503b3a1-7d03-41b2-9ec8-e478c5414d67  79fd602b-3e8e-4353-aa78-4f5f170b607d  \\\n",
       "0                                  4884                                  2830   \n",
       "1                                     2                                     3   \n",
       "2                                  5054                                  1045   \n",
       "3                                   890                                   718   \n",
       "4                                   626                                   198   \n",
       "\n",
       "   07a859fc-6f78-4905-9035-90e2403dbe8d  f9c835db-2ab6-4bf5-826f-48723493c0ec  \\\n",
       "0                                  5286                                  3635   \n",
       "1                                     4                                     3   \n",
       "2                                  3846                                  5822   \n",
       "3                                   981                                  1033   \n",
       "4                                   753                                   915   \n",
       "\n",
       "   a782352e-c4bb-4c3e-ba82-456a47c3689a  \n",
       "0                                  4420  \n",
       "1                                     0  \n",
       "2                                  4038  \n",
       "3                                   978  \n",
       "4                                   724  \n",
       "\n",
       "[5 rows x 507 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_symbol = pd.read_csv(\"data.csv\")\n",
    "df_symbol.head()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Clinical Data\n",
    "---"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['diagnoses.gleason_patterns_percent', 'diagnoses.synchronous_malignancy', 'diagnoses.child_pugh_classification', 'diagnoses.submitter_id', 'diagnoses.cog_neuroblastoma_risk_group', 'diagnoses.peritoneal_fluid_cytological_status', 'diagnoses.cog_renal_stage', 'diagnoses.esophageal_columnar_metaplasia_present', 'diagnoses.figo_staging_edition_year', 'diagnoses.primary_diagnosis', 'diagnoses.circumferential_resection_margin', 'diagnoses.diagnosis_id', 'diagnoses.tumor_largest_dimension_diameter', 'diagnoses.papillary_renal_cell_type', 'diagnoses.primary_gleason_grade', 'diagnoses.pregnant_at_diagnosis', 'diagnoses.ajcc_clinical_n', 'diagnoses.site_of_resection_or_biopsy', 'diagnoses.lymph_nodes_positive', 'diagnoses.days_to_recurrence', 'diagnoses.laterality', 'diagnoses.non_nodal_regional_disease', 'diagnoses.irs_group', 'diagnoses.medulloblastoma_molecular_classification', 'diagnoses.masaoka_stage', 'diagnoses.micropapillary_features', 'diagnoses.ann_arbor_b_symptoms', 'diagnoses.best_overall_response', 'diagnoses.days_to_last_known_disease_status', 'diagnoses.ovarian_surface_involvement', 'diagnoses.progression_or_recurrence', 'diagnoses.vascular_invasion_present', 'diagnoses.year_of_diagnosis', 'diagnoses.ajcc_pathologic_n', 'diagnoses.first_symptom_prior_to_diagnosis', 'diagnoses.prior_malignancy', 'diagnoses.international_prognostic_index', 'diagnoses.cog_liver_stage', 'diagnoses.tumor_grade', 'diagnoses.anaplasia_present_type', 'diagnoses.classification_of_tumor', 'diagnoses.greatest_tumor_dimension', 'diagnoses.enneking_msts_tumor_site', 'diagnoses.esophageal_columnar_dysplasia_degree', 'diagnoses.gastric_esophageal_junction_involvement', 'diagnoses.tumor_stage', 'diagnoses.age_at_diagnosis', 'diagnoses.non_nodal_tumor_deposits', 'diagnoses.tumor_depth', 'diagnoses.cog_rhabdomyosarcoma_risk_group', 'diagnoses.treatments.treatment_effect', 'diagnoses.treatments.created_datetime', 'diagnoses.treatments.treatment_id', 'diagnoses.treatments.submitter_id', 'diagnoses.treatments.days_to_treatment_start', 'diagnoses.treatments.therapeutic_agents', 'diagnoses.treatments.days_to_treatment_end', 'diagnoses.treatments.regimen_or_line_of_therapy', 'diagnoses.treatments.state', 'diagnoses.treatments.treatment_outcome', 'diagnoses.treatments.reason_treatment_ended', 'diagnoses.treatments.treatment_arm', 'diagnoses.treatments.treatment_frequency', 'diagnoses.treatments.treatment_anatomic_site', 'diagnoses.treatments.treatment_type', 'diagnoses.treatments.chemo_concurrent_to_radiation', 'diagnoses.treatments.treatment_effect_indicator', 'diagnoses.treatments.treatment_intent_type', 'diagnoses.treatments.initial_disease_status', 'diagnoses.treatments.treatment_dose', 'diagnoses.treatments.treatment_or_therapy', 'diagnoses.treatments.number_of_cycles', 'diagnoses.treatments.updated_datetime', 'diagnoses.treatments.treatment_dose_units', 'diagnoses.weiss_assessment_score', 'diagnoses.largest_extrapelvic_peritoneal_focus', 'diagnoses.irs_stage', 'diagnoses.margin_distance', 'diagnoses.inss_stage', 'diagnoses.ajcc_pathologic_t', 'diagnoses.created_datetime', 'diagnoses.margins_involved_site', 'diagnoses.ajcc_clinical_m', 'diagnoses.lymphatic_invasion_present', 'diagnoses.prior_treatment', 'diagnoses.ajcc_clinical_t', 'diagnoses.inrg_stage', 'diagnoses.tumor_focality', 'diagnoses.state', 'diagnoses.ann_arbor_extranodal_involvement', 'diagnoses.inpc_grade', 'diagnoses.iss_stage', 'diagnoses.last_known_disease_status', 'diagnoses.transglottic_extension', 'diagnoses.mitotic_count', 'diagnoses.tissue_or_organ_of_origin', 'diagnoses.anaplasia_present', 'diagnoses.supratentorial_localization', 'diagnoses.figo_stage', 'diagnoses.ann_arbor_clinical_stage', 'diagnoses.days_to_best_overall_response', 'diagnoses.morphology', 'diagnoses.enneking_msts_metastasis', 'diagnoses.gleason_grade_tertiary', 'diagnoses.metastasis_at_diagnosis', 'diagnoses.ann_arbor_pathologic_stage', 'diagnoses.ajcc_staging_system_edition', 'diagnoses.peripancreatic_lymph_nodes_positive', 'diagnoses.method_of_diagnosis', 'diagnoses.lymph_nodes_tested', 'diagnoses.wilms_tumor_histologic_subtype', 'diagnoses.breslow_thickness', 'diagnoses.ajcc_pathologic_stage', 'diagnoses.lymph_node_involved_site', 'diagnoses.mitosis_karyorrhexis_index', 'diagnoses.gleason_grade_group', 'diagnoses.gross_tumor_weight', 'diagnoses.burkitt_lymphoma_clinical_variant', 'diagnoses.ajcc_clinical_stage', 'diagnoses.ovarian_specimen_status', 'diagnoses.peripancreatic_lymph_nodes_tested', 'diagnoses.enneking_msts_grade', 'diagnoses.igcccg_stage', 'diagnoses.annotations.created_datetime', 'diagnoses.annotations.annotation_id', 'diagnoses.annotations.submitter_id', 'diagnoses.annotations.status', 'diagnoses.annotations.updated_datetime', 'diagnoses.annotations.case_submitter_id', 'diagnoses.annotations.entity_submitter_id', 'diagnoses.annotations.entity_id', 'diagnoses.annotations.state', 'diagnoses.annotations.notes', 'diagnoses.annotations.legacy_created_datetime', 'diagnoses.annotations.legacy_updated_datetime', 'diagnoses.annotations.classification', 'diagnoses.annotations.creator', 'diagnoses.annotations.case_id', 'diagnoses.annotations.entity_type', 'diagnoses.annotations.category', 'diagnoses.vascular_invasion_type', 'diagnoses.ishak_fibrosis_score', 'diagnoses.metastasis_at_diagnosis_site', 'diagnoses.ajcc_pathologic_m', 'diagnoses.updated_datetime', 'diagnoses.days_to_last_follow_up', 'diagnoses.enneking_msts_stage', 'diagnoses.goblet_cells_columnar_mucosa_present', 'diagnoses.secondary_gleason_grade', 'diagnoses.tumor_confined_to_organ_of_origin', 'diagnoses.days_to_diagnosis', 'diagnoses.inpc_histologic_group', 'diagnoses.percent_tumor_invasion', 'diagnoses.tumor_regression_grade', 'diagnoses.icd_10_code', 'diagnoses.residual_disease', 'diagnoses.perineural_invasion_present']\n",
      "['demographic.vital_status', 'demographic.premature_at_birth', 'demographic.created_datetime', 'demographic.cause_of_death', 'demographic.submitter_id', 'demographic.age_is_obfuscated', 'demographic.updated_datetime', 'demographic.year_of_death', 'demographic.race', 'demographic.demographic_id', 'demographic.country_of_residence_at_enrollment', 'demographic.days_to_death', 'demographic.year_of_birth', 'demographic.gender', 'demographic.days_to_birth', 'demographic.cause_of_death_source', 'demographic.ethnicity', 'demographic.weeks_gestation_at_birth', 'demographic.occupation_duration_years', 'demographic.state', 'demographic.age_at_index']\n"
     ]
    }
   ],
   "source": [
    "# get all demographic fields\n",
    "\n",
    "cases_fields = requests.get(cases_endpt + \"_mapping\").json()[\"fields\"]\n",
    "\n",
    "diagnoses_fields = list(filter(lambda x: \"diagnoses\" in x, cases_fields))\n",
    "print(diagnoses_fields)\n",
    "\n",
    "demographic_fields = list(filter(lambda x: \"demographic\" in x, cases_fields))\n",
    "print(demographic_fields)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "# initialize dataframe\n",
    "demographic_column_names = [ field.split(\".\")[1] for field in demographic_fields ]\n",
    "diagnoses_column_names = [ field.split(\".\")[1] for field in diagnoses_fields ]\n",
    "\n",
    "columns = list(set([*demographic_column_names,*diagnoses_column_names]))\n",
    "df_clinical = pd.DataFrame({}, columns=columns)\n",
    "\n",
    "df_clinical[\"case_id\"] = []\n",
    "\n",
    "# get demographics and diagnosis data for each case, \n",
    "# merging with pre-exisiting dataframe\n",
    "\n",
    "for case in case_uuid_list:\n",
    "    fields=\",\".join([*demographic_fields, *diagnoses_fields])\n",
    "    params={\n",
    "        \"fields\": fields\n",
    "    }\n",
    "    response = requests.get(cases_endpt + case, params=params).json()[\"data\"]\n",
    "    demographic_data = response[\"demographic\"]\n",
    "    diagnoses_data = response[\"diagnoses\"]\n",
    "    diagnoses_data = diagnoses_data[0]\n",
    "    del diagnoses_data[\"treatments\"] # do not load treatment data\n",
    "    df_case = pd.DataFrame({**demographic_data,**diagnoses_data}, index=[case])\n",
    "    df_case.head()\n",
    "    df_case[\"case_id\"] = case\n",
    "    df_clinical = pd.concat([df_clinical, df_case], join=\"outer\")\n",
    "\n",
    "df_clinical = df_clinical.set_index(\"case_id\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>primary_diagnosis</th>\n",
       "      <th>primary_gleason_grade</th>\n",
       "      <th>figo_staging_edition_year</th>\n",
       "      <th>updated_datetime</th>\n",
       "      <th>occupation_duration_years</th>\n",
       "      <th>ann_arbor_clinical_stage</th>\n",
       "      <th>best_overall_response</th>\n",
       "      <th>largest_extrapelvic_peritoneal_focus</th>\n",
       "      <th>esophageal_columnar_dysplasia_degree</th>\n",
       "      <th>perineural_invasion_present</th>\n",
       "      <th>...</th>\n",
       "      <th>lymph_node_involved_site</th>\n",
       "      <th>year_of_diagnosis</th>\n",
       "      <th>ann_arbor_b_symptoms</th>\n",
       "      <th>gastric_esophageal_junction_involvement</th>\n",
       "      <th>days_to_birth</th>\n",
       "      <th>ajcc_clinical_stage</th>\n",
       "      <th>treatments</th>\n",
       "      <th>ajcc_pathologic_stage</th>\n",
       "      <th>progression_or_recurrence</th>\n",
       "      <th>ajcc_pathologic_t</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>case_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>d79e692c-5053-4484-a180-01a094c5ff45</th>\n",
       "      <td>Serous cystadenocarcinoma, NOS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2019-08-27T13:09:24.298212-05:00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2010</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-25612</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not reported</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32b5d0e4-a242-44d1-ac20-6e5ca2f35b71</th>\n",
       "      <td>Serous cystadenocarcinoma, NOS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2019-08-27T13:11:28.180392-05:00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2007</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-23913</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not reported</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16e69011-c295-479f-b521-86e66fba498d</th>\n",
       "      <td>Serous cystadenocarcinoma, NOS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2019-08-27T13:10:55.555572-05:00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2008</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-27065</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not reported</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>211b2c52-1ad8-4d33-af0b-69ec1a56c821</th>\n",
       "      <td>Serous cystadenocarcinoma, NOS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2019-08-27T13:08:19.563638-05:00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2010</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-20594</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not reported</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37d5c1d9-3ff5-44fe-afdc-ff430a0f3510</th>\n",
       "      <td>Serous cystadenocarcinoma, NOS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2019-08-27T13:10:55.555572-05:00</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2010</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-28037</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>not reported</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 135 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   primary_diagnosis  \\\n",
       "case_id                                                                \n",
       "d79e692c-5053-4484-a180-01a094c5ff45  Serous cystadenocarcinoma, NOS   \n",
       "32b5d0e4-a242-44d1-ac20-6e5ca2f35b71  Serous cystadenocarcinoma, NOS   \n",
       "16e69011-c295-479f-b521-86e66fba498d  Serous cystadenocarcinoma, NOS   \n",
       "211b2c52-1ad8-4d33-af0b-69ec1a56c821  Serous cystadenocarcinoma, NOS   \n",
       "37d5c1d9-3ff5-44fe-afdc-ff430a0f3510  Serous cystadenocarcinoma, NOS   \n",
       "\n",
       "                                     primary_gleason_grade  \\\n",
       "case_id                                                      \n",
       "d79e692c-5053-4484-a180-01a094c5ff45                   NaN   \n",
       "32b5d0e4-a242-44d1-ac20-6e5ca2f35b71                   NaN   \n",
       "16e69011-c295-479f-b521-86e66fba498d                   NaN   \n",
       "211b2c52-1ad8-4d33-af0b-69ec1a56c821                   NaN   \n",
       "37d5c1d9-3ff5-44fe-afdc-ff430a0f3510                   NaN   \n",
       "\n",
       "                                     figo_staging_edition_year  \\\n",
       "case_id                                                          \n",
       "d79e692c-5053-4484-a180-01a094c5ff45                       NaN   \n",
       "32b5d0e4-a242-44d1-ac20-6e5ca2f35b71                       NaN   \n",
       "16e69011-c295-479f-b521-86e66fba498d                       NaN   \n",
       "211b2c52-1ad8-4d33-af0b-69ec1a56c821                       NaN   \n",
       "37d5c1d9-3ff5-44fe-afdc-ff430a0f3510                       NaN   \n",
       "\n",
       "                                                      updated_datetime  \\\n",
       "case_id                                                                  \n",
       "d79e692c-5053-4484-a180-01a094c5ff45  2019-08-27T13:09:24.298212-05:00   \n",
       "32b5d0e4-a242-44d1-ac20-6e5ca2f35b71  2019-08-27T13:11:28.180392-05:00   \n",
       "16e69011-c295-479f-b521-86e66fba498d  2019-08-27T13:10:55.555572-05:00   \n",
       "211b2c52-1ad8-4d33-af0b-69ec1a56c821  2019-08-27T13:08:19.563638-05:00   \n",
       "37d5c1d9-3ff5-44fe-afdc-ff430a0f3510  2019-08-27T13:10:55.555572-05:00   \n",
       "\n",
       "                                     occupation_duration_years  \\\n",
       "case_id                                                          \n",
       "d79e692c-5053-4484-a180-01a094c5ff45                       NaN   \n",
       "32b5d0e4-a242-44d1-ac20-6e5ca2f35b71                       NaN   \n",
       "16e69011-c295-479f-b521-86e66fba498d                       NaN   \n",
       "211b2c52-1ad8-4d33-af0b-69ec1a56c821                       NaN   \n",
       "37d5c1d9-3ff5-44fe-afdc-ff430a0f3510                       NaN   \n",
       "\n",
       "                                     ann_arbor_clinical_stage  \\\n",
       "case_id                                                         \n",
       "d79e692c-5053-4484-a180-01a094c5ff45                      NaN   \n",
       "32b5d0e4-a242-44d1-ac20-6e5ca2f35b71                      NaN   \n",
       "16e69011-c295-479f-b521-86e66fba498d                      NaN   \n",
       "211b2c52-1ad8-4d33-af0b-69ec1a56c821                      NaN   \n",
       "37d5c1d9-3ff5-44fe-afdc-ff430a0f3510                      NaN   \n",
       "\n",
       "                                     best_overall_response  \\\n",
       "case_id                                                      \n",
       "d79e692c-5053-4484-a180-01a094c5ff45                   NaN   \n",
       "32b5d0e4-a242-44d1-ac20-6e5ca2f35b71                   NaN   \n",
       "16e69011-c295-479f-b521-86e66fba498d                   NaN   \n",
       "211b2c52-1ad8-4d33-af0b-69ec1a56c821                   NaN   \n",
       "37d5c1d9-3ff5-44fe-afdc-ff430a0f3510                   NaN   \n",
       "\n",
       "                                     largest_extrapelvic_peritoneal_focus  \\\n",
       "case_id                                                                     \n",
       "d79e692c-5053-4484-a180-01a094c5ff45                                  NaN   \n",
       "32b5d0e4-a242-44d1-ac20-6e5ca2f35b71                                  NaN   \n",
       "16e69011-c295-479f-b521-86e66fba498d                                  NaN   \n",
       "211b2c52-1ad8-4d33-af0b-69ec1a56c821                                  NaN   \n",
       "37d5c1d9-3ff5-44fe-afdc-ff430a0f3510                                  NaN   \n",
       "\n",
       "                                     esophageal_columnar_dysplasia_degree  \\\n",
       "case_id                                                                     \n",
       "d79e692c-5053-4484-a180-01a094c5ff45                                  NaN   \n",
       "32b5d0e4-a242-44d1-ac20-6e5ca2f35b71                                  NaN   \n",
       "16e69011-c295-479f-b521-86e66fba498d                                  NaN   \n",
       "211b2c52-1ad8-4d33-af0b-69ec1a56c821                                  NaN   \n",
       "37d5c1d9-3ff5-44fe-afdc-ff430a0f3510                                  NaN   \n",
       "\n",
       "                                     perineural_invasion_present  ...  \\\n",
       "case_id                                                           ...   \n",
       "d79e692c-5053-4484-a180-01a094c5ff45                         NaN  ...   \n",
       "32b5d0e4-a242-44d1-ac20-6e5ca2f35b71                         NaN  ...   \n",
       "16e69011-c295-479f-b521-86e66fba498d                         NaN  ...   \n",
       "211b2c52-1ad8-4d33-af0b-69ec1a56c821                         NaN  ...   \n",
       "37d5c1d9-3ff5-44fe-afdc-ff430a0f3510                         NaN  ...   \n",
       "\n",
       "                                     lymph_node_involved_site  \\\n",
       "case_id                                                         \n",
       "d79e692c-5053-4484-a180-01a094c5ff45                      NaN   \n",
       "32b5d0e4-a242-44d1-ac20-6e5ca2f35b71                      NaN   \n",
       "16e69011-c295-479f-b521-86e66fba498d                      NaN   \n",
       "211b2c52-1ad8-4d33-af0b-69ec1a56c821                      NaN   \n",
       "37d5c1d9-3ff5-44fe-afdc-ff430a0f3510                      NaN   \n",
       "\n",
       "                                     year_of_diagnosis ann_arbor_b_symptoms  \\\n",
       "case_id                                                                       \n",
       "d79e692c-5053-4484-a180-01a094c5ff45              2010                  NaN   \n",
       "32b5d0e4-a242-44d1-ac20-6e5ca2f35b71              2007                  NaN   \n",
       "16e69011-c295-479f-b521-86e66fba498d              2008                  NaN   \n",
       "211b2c52-1ad8-4d33-af0b-69ec1a56c821              2010                  NaN   \n",
       "37d5c1d9-3ff5-44fe-afdc-ff430a0f3510              2010                  NaN   \n",
       "\n",
       "                                     gastric_esophageal_junction_involvement  \\\n",
       "case_id                                                                        \n",
       "d79e692c-5053-4484-a180-01a094c5ff45                                     NaN   \n",
       "32b5d0e4-a242-44d1-ac20-6e5ca2f35b71                                     NaN   \n",
       "16e69011-c295-479f-b521-86e66fba498d                                     NaN   \n",
       "211b2c52-1ad8-4d33-af0b-69ec1a56c821                                     NaN   \n",
       "37d5c1d9-3ff5-44fe-afdc-ff430a0f3510                                     NaN   \n",
       "\n",
       "                                     days_to_birth ajcc_clinical_stage  \\\n",
       "case_id                                                                  \n",
       "d79e692c-5053-4484-a180-01a094c5ff45        -25612                 NaN   \n",
       "32b5d0e4-a242-44d1-ac20-6e5ca2f35b71        -23913                 NaN   \n",
       "16e69011-c295-479f-b521-86e66fba498d        -27065                 NaN   \n",
       "211b2c52-1ad8-4d33-af0b-69ec1a56c821        -20594                 NaN   \n",
       "37d5c1d9-3ff5-44fe-afdc-ff430a0f3510        -28037                 NaN   \n",
       "\n",
       "                                     treatments ajcc_pathologic_stage  \\\n",
       "case_id                                                                 \n",
       "d79e692c-5053-4484-a180-01a094c5ff45        NaN                   NaN   \n",
       "32b5d0e4-a242-44d1-ac20-6e5ca2f35b71        NaN                   NaN   \n",
       "16e69011-c295-479f-b521-86e66fba498d        NaN                   NaN   \n",
       "211b2c52-1ad8-4d33-af0b-69ec1a56c821        NaN                   NaN   \n",
       "37d5c1d9-3ff5-44fe-afdc-ff430a0f3510        NaN                   NaN   \n",
       "\n",
       "                                     progression_or_recurrence  \\\n",
       "case_id                                                          \n",
       "d79e692c-5053-4484-a180-01a094c5ff45              not reported   \n",
       "32b5d0e4-a242-44d1-ac20-6e5ca2f35b71              not reported   \n",
       "16e69011-c295-479f-b521-86e66fba498d              not reported   \n",
       "211b2c52-1ad8-4d33-af0b-69ec1a56c821              not reported   \n",
       "37d5c1d9-3ff5-44fe-afdc-ff430a0f3510              not reported   \n",
       "\n",
       "                                     ajcc_pathologic_t  \n",
       "case_id                                                 \n",
       "d79e692c-5053-4484-a180-01a094c5ff45               NaN  \n",
       "32b5d0e4-a242-44d1-ac20-6e5ca2f35b71               NaN  \n",
       "16e69011-c295-479f-b521-86e66fba498d               NaN  \n",
       "211b2c52-1ad8-4d33-af0b-69ec1a56c821               NaN  \n",
       "37d5c1d9-3ff5-44fe-afdc-ff430a0f3510               NaN  \n",
       "\n",
       "[5 rows x 135 columns]"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# make first column \"primary_diagnosis\"\n",
    "cols = ['primary_diagnosis']  + [col for col in df_clinical.columns.values if col != 'primary_diagnosis']\n",
    "df_clinical = df_clinical[cols]\n",
    "df_clinical.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save final .csv\n",
    "df_clinical.to_csv(\"clinical_data.csv\", encoding='utf-8')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save a df containing all entries for one cancer type for each type\n",
    "cancer_types = list(set(df_clinical.primary_diagnosis))\n",
    "\n",
    "for cancer in cancer_types:\n",
    "    df_cancer = df_clinical[df_clinical.primary_diagnosis == cancer]\n",
    "    df_cancer.to_csv(\"data_cancer/{}.csv\".format(cancer), encoding='utf-8')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "TCGA",
   "language": "python",
   "name": "tcga"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
